File: src/main/java/com/nibado/projects/advent/Day11.java
package com.nibado.projects.advent;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResourceAsString;

public class Day11 implements Runnable {
    private static final char[] ILLEGAL = "ilo".toCharArray();
    @Override
    public void run() {
        char[] password = readResourceAsString("/day11.txt").toCharArray();

        nextValid(password);
        printAnswer(11, "One", new String(password));
        nextValid(password);
        printAnswer(11, "Two", new String(password));
    }

    public static void nextValid(char[] password) {
        do {
            increment(password);
        }
        while(!hasStraight(password, 3) || containsIllegal(password) || countPairs(password) < 2);
    }

    public static void increment(char[] password) {
        for(int i = password.length - 1;i >= 0;i--) {
            password[i]++;
            if(password[i] > 'z') {
                password[i] = 'a';
            }
            else {
                break;
            }
        }
    }

    public static boolean hasStraight(char[] password, int length) {
        for(int i = 0;i < password.length - length;i++) {
            int count = 0;
            for(int j = 0;j < length;j++) {
                if(password[i + j] == password[i] + (char)j) {
                    count ++;
                }
            }
            if(count == length) {
                return true;
            }
        }

        return false;
    }

    public static boolean containsIllegal(char[] password) {
        for(int i = 0;i < password.length;i++) {
            for(int j = 0;j < ILLEGAL.length;j++) {
                if(password[i] == ILLEGAL[j]) {
                    return true;
                }
            }
        }

        return false;
    }

    public static int countPairs(char[] password) {
        int count = 0;
        for(int i = 0;i < password.length - 1;i++) {
            if(password[i] == password[i + 1]) {
                count++;
                i++;
            }
        }

        return count;
    }

    public static void main(String... argv) {
        new Day11().run();
    }
}

File: src/main/scala/com/nibado/projects/adventscala/Day11.scala
package com.nibado.projects.adventscala

class Day11 extends Day {
  val illegal = "ilo".toList

  override def run(): Unit = {
    val password = new StringBuilder(getResourceAsString("/day11.txt"))

    printAnswer(11, "One", nextValid(password))
    printAnswer(11, "Two", nextValid(password))
  }

  def increment(password: StringBuilder): Unit = {
    (password.length - 1 to 0 by -1).foreach(i => {
      password.update(i, (password(i) + 1).toChar)
      if(password(i) > 'z') {
        password.update(i, 'a')
      }
      else {
        return
      }
    })
  }

  def nextValid(password: StringBuilder): StringBuilder = {
    do {
      increment(password)
    }
    while(!threeStraight(password) || containsIllegal(password) || !hasTwoPairs(password))

    password
  }

  def threeStraight(s: StringBuilder) =
    s.toList.sliding(3).exists { case Seq(a, b, c) =>
      a - b == -1 && b - c == -1
    }

  def containsIllegal(s: StringBuilder) =
    s.exists(c => illegal.contains(c))

  def hasTwoPairs(s: StringBuilder) = {
    val pairs = s.sliding(2).toSeq

    pairs.lastIndexWhere(l => l(0) == l(1)) - pairs.indexWhere(l => l(0) == l(1)) > 1
  }
}
